rlapy/comps/evd.py

```diff
@@ -0,0 +1,432 @@
+#from rlapy.comps.svd import SVDecomposer, SVD1
+import rlapy.utils.linalg_wrappers as ulaw
+
+###############################################################################
+#       Classic implementations, exposing fewest possible parameters.
+###############################################################################
+
+
+def evd1(num_passes, A, k, epsilon, s, rng):
+    """
+    Return the eigen decomposition matrices (V, lambda_matrix),
+    based on a rank-k QB factorization of A.
+    Use a Gaussian sketching matrix and pass over A a total of
+    num_passes times.
+
+    Parameters
+    ----------
+    num_passes : int
+        Total number of passes over A. We require num_passes >= 2.
+
+    A : Union[ndarray, spmatrix, LinearOperator]
+        Data matrix to approximate. A must be an n × n Hermitian matrix.
+
+    k : int
+        Target rank for the approximation of A: 0 < k < min(A.shape).
+        This parameter includes any oversampling. For example, if you
+        want to be near the optimal (Eckhart-Young) error for a rank 20
+        approximation of A, then you might want to set k=25.
+        
+    epsilon : float
+        Target accuracy for the oversampled approximation of A: 0 < epsilon < np.inf.
+        This parameter inherits from the QBDecomposer or RangeFinder class.
+        
+    s  : int
+        Auxiliary parameter for the QBDecomposer or RangeFinder.
+
+    rng : Union[None, int, SeedSequence, BitGenerator, Generator]
+        Determines the numpy Generator object that manages randomness
+        in this function call.
+
+    Returns
+    -------
+    V : ndarray
+        Has shape (A.shape[0], k). Columns are orthonormal.
+
+    lambda_matrix : ndarray
+        Has shape (k,k). lambda_matrix is a diagonal matrix storing the eigenvalues of 
+        the matrix A
+
+    Notes
+    -----
+    We perform (num_passes - 2) steps of subspace iteration, and
+    stabilize subspace iteration by a QR factorization at every step.
+
+    References
+    ----------
+    This algorithm computes V and then sets B = Q.T @ A. Conceptually, we
+    compute Q by adapting Algorithm 5.3 from
+
+        Halko, Nathan, Per-Gunnar Martinsson, and Joel A. Tropp.
+        "Finding structure with randomness: Probabilistic algorithms for
+        constructing approximate matrix decompositions."
+        SIAM review 53.2 (2011): 217-288.
+        (available at `arXiv <http://arxiv.org/abs/0909.4061>`_).
+    EVDecomposer: adapted from the QB-based [HMT11, Algorithm 5.3]
+    """
+    if not (A == A.T).all():
+        #Needs to consider conjugate as well if A contains complex numbers.
+        msg = """
+        This rountine evd1 only works for Hermitian matrices.
+        """ 
+        raise RuntimeError(msg)
+    rng = np.random.default_rng(rng)
+    rso_ = RS1(gaussian_operator, num_passes - 2, ulaw.orth, 1)
+    rf_ = RF1(rso_)
+    qb_ = QB1(rf_)
+    #We need QB1 not QB2, since B=Q^*A is necessary
```

Comment:
The ``QBDecomposer`` interface requires that the returned matrices ``(Q, B)`` satisfy ``B = Q.T @ A``. So you can use any QB method.